internal/pkg/term/syncbuffer/syncbuffer.go

```diff
@@ -0,0 +1,112 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package syncbuffer provides a goroutine safe bytes.Buffer as well printing functionality to the terminal.
+package syncbuffer
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+	"sync"
+)
+
+// fileReader interface that can read data from a file.
+// It extends the io.Reader interface, which provides a method for reading bytes into a buffer.
+type fileReader interface {
+	io.Reader
+}
+
+// SyncBuffer is a synchronized buffer that can be used to store output data and coordinate between multiple goroutines.
+type SyncBuffer struct {
+	bufMu sync.Mutex    // bufMu is a mutex that can be used to protect access to the buffer.
+	buf   bytes.Buffer  // buf is the buffer that stores the data.
+	done  chan struct{} // done is a channel that can be used to signal when the operations are complete.
+}
+
+// NewSyncBuffer creates and returns a new SyncBuffer object with an initialized 'done' channel.
+func NewSyncBuffer() *SyncBuffer {
+	return &SyncBuffer{
+		done: make(chan struct{}),
+	}
+}
+
+// Write appends the given bytes to the buffer.
+func (b *SyncBuffer) Write(p []byte) (n int, err error) {
+	b.bufMu.Lock()
+	defer b.bufMu.Unlock()
+	return b.buf.Write(p)
+}
+
+// strings returns an empty slice if the buffer is empty.
+// Otherwise, it returns a slice of all the lines stored in the buffer.
+func (b *SyncBuffer) strings() []string {
+	b.bufMu.Lock()
+	defer b.bufMu.Unlock()
+	lines := b.buf.String()
+	if len(lines) == 0 {
+		return nil
+	}
+	return strings.Split(strings.TrimSpace(lines), "\n")
+}
+
+// IsDone returns true if the Done channel has been closed, otherwise return false.
+func (b *SyncBuffer) IsDone() bool {
+	select {
+	case <-b.done:
+		return true
+	default:
+		return false
+	}
+}
+
+// MarkDone closes the Done channel.
+func (b *SyncBuffer) MarkDone() {
+	close(b.done)
+}
+
+// CopyToBuffer reads from a fileReader and writes the data to the buffer.
+func (b *SyncBuffer) CopyToBuffer(r fileReader) error {
```

Comment:
I wonder if this should be another constructor, besides `New`, like:
```suggestion
// Copy reads all the content of an io.Reader into a SyncBuffer and returns it.
func Copy(r io.Reader) (*SyncBuffer, error) {
```